d-sandbox

<div style="text-align: center; line-height: 0; padding-top: 9px;">
  <img src="https://databricks.com/wp-content/uploads/2018/03/db-academy-rgb-1200px.png" alt="Databricks Learning" style="width: 600px">
</div>

# Delta Lake
## Module 4 Assignment

This final assignment is broken up into 2 parts:
1. Completing this Delta Lake notebook
  * Submitting question answers to Coursera
  * Uploading notebook to Coursera for peer reviewing
2. Answering 3 free response questions on Coursera platform

## ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) In this notebook you:
* Create a Delta table with partitioned columns
* Delete records
* Update existing records
* Time travel!


For each **bold** question, input its answer in Coursera.

In [0]:
%run ../Includes/Classroom-Setup

Create a new Delta table called `fireCallsPartitioned` partitioned by `Priority`.

In [0]:
%sql
CREATE OR REPLACE TEMPORARY VIEW fireCallsParquet
USING parquet
OPTIONS (
  path "/mnt/davis/fire-calls/fire-calls-clean.parquet/"
)

In [0]:
%sql
-- ANSWER
CREATE DATABASE IF NOT EXISTS Databricks;
USE Databricks;
DROP TABLE IF EXISTS fireCallsPartitioned;

CREATE TABLE fireCallsPartitioned

USING DELTA
PARTITIONED BY (Priority)
AS 
  SELECT * FROM fireCallsParquet

num_affected_rows num_inserted_rows

### Question 1

**How many folders were created? Enter the number of records you see from the output below (include the `_delta_log` in your count).**<br>
**9 Including the _delata log**

In [0]:
%fs ls dbfs:/user/hive/warehouse/databricks.db/firecallspartitioned

path name size dbfs:/user/hive/warehouse/databricks.db/firecallspartitioned/Priority=1/ Priority=1/ 0 dbfs:/user/hive/warehouse/databricks.db/firecallspartitioned/Priority=2/ Priority=2/ 0 dbfs:/user/hive/warehouse/databricks.db/firecallspartitioned/Priority=3/ Priority=3/ 0 dbfs:/user/hive/warehouse/databricks.db/firecallspartitioned/Priority=A/ Priority=A/ 0 dbfs:/user/hive/warehouse/databricks.db/firecallspartitioned/Priority=B/ Priority=B/ 0 dbfs:/user/hive/warehouse/databricks.db/firecallspartitioned/Priority=C/ Priority=C/ 0 dbfs:/user/hive/warehouse/databricks.db/firecallspartitioned/Priority=E/ Priority=E/ 0 dbfs:/user/hive/warehouse/databricks.db/firecallspartitioned/Priority=I/ Priority=I/ 0 dbfs:/user/hive/warehouse/databricks.db/firecallspartitioned/_delta_log/ _delta_log/ 0

### Question 2

**Delete all the records where `City` is null. How many records are left in the delta table?** **550**

In [0]:
%sql
select count(*) from firecallspartitioned  where city is null 

count(1) 550

In [0]:
%sql
-- ANSWER
DELETE FROM firecallspartitioned WHERE City is null

num_affected_rows 550

In [0]:
%sql
select count(*) from firecallspartitioned

count(1) 416869

### Question 3

**After you deleted all records where the City is null, how many files were removed? Hint: Look at `operationsMetrics` in the transaction log using the `DESCRIBE HISTORY` command.**<br>
**22**

In [0]:
%sql
-- ANSWER
DESCRIBE HISTORY firecallspartitioned

version timestamp userId userName operation operationParameters job notebook clusterId readVersion isolationLevel isBlindAppend operationMetrics userMetadata 1 2022-02-12T13:41:38.000+0000 5779868209403466 birhan.hailu.tigray@gmail.com DELETE Map(predicate -> ["(spark_catalog.databricks.firecallspartitioned.`City` IS NULL)"]) null List(676651046033364) 0212-123705-nxew3a5b 0 WriteSerializable false Map(numRemovedFiles -> 22, numCopiedRows -> 387067, numAddedChangeFiles -> 0, executionTimeMs -> 36173, numDeletedRows -> 550, scanTimeMs -> 6725, numAddedFiles -> 9, rewriteTimeMs -> 29447) null 0 2022-02-12T13:40:47.000+0000 5779868209403466 birhan.hailu.tigray@gmail.com CREATE TABLE AS SELECT Map(isManaged -> true, description -> null, partitionBy -> ["Priority"], properties -> {}) null List(676651046033364) 0212-123705-nxew3a5b null WriteSerializable true Map(numFiles -> 62, numOutputBytes -> 53387546, numOutputRows -> 417419) null

### Question 4

**There are quite a few missing `Call_Type_Group` values. Use the `UPDATE` command to replace any null values with `Non Life-threatening`.**

**After you replace the null values, how many `Non Life-threatening` call types are there?** 246459

In [0]:
%sql
-- ANSWER 
UPDATE firecallspartitioned SET Call_Type_Group="Non Life-threatening" WHERE Call_Type_Group is null

num_affected_rows 246459

### Question 5

**Travel back in time to the earliest version of the Delta table (version 0). How many records were there?** **417419**

In [0]:
%sql
-- ANSWER 

SELECT COUNT(*)
FROM firecallspartitioned 
  VERSION AS OF 0 

count(1) 417419

Congrats on finishing your last assignment notebook! 


Now you will have to upload this notebook to Coursera for peer reviewing.
1. Make sure that all your code will run without errors
  * Check this by clicking the "Clear State & Run All" dropdown option at the top of your notebook
  * ![](http://files.training.databricks.com/images/eLearning/ucdavis/clearstaterunall.png)
2. Click on the "Workspace" icon on the side bar
3. Next to the notebook you're working in right now, click on the dropdown arrow
4. In the dropdown, click on "Export" then "HTML"
  * ![](http://files.training.databricks.com/images/davis/upload_html_screenshot.png)
5. On the Coursera platform, upload this HTML file to Week 4's Peer Review Assignment

Go back onto the Coursera platform for the free response portion of this assignment and for instructions on how to review your peer's work.

-sandbox
&copy; 2021 Databricks, Inc. All rights reserved.<br/>
Apache, Apache Spark, Spark and the Spark logo are trademarks of the <a href="http://www.apache.org/">Apache Software Foundation</a>.<br/>
<br/>
<a href="https://databricks.com/privacy-policy">Privacy Policy</a> | <a href="https://databricks.com/terms-of-use">Terms of Use</a> | <a href="http://help.databricks.com/">Support</a>